### PR TITLE
Added a decorator to connect to the heat pump

### LINKED ISSUE #130: Get rid of _read_after_write()
The connection to the heat pump is currently being established within the `_read_after_write()` function.

This makes it difficult to offer additional functions such as `read_parameters()`.

### PR DESCRIPTION
**Pull request by Guzz-T**
Using a decorator we can simply re-use the code to connect to the heat pump. This allows us to offers several read / write functions like only read the paramters.

Fixes #130

### CODE REVIEW

**Comment by gerw** [comment]
Shouldn't it be possible to use your `_connect` as a "real" decorator (with `@`) around the read/write methods?

**Comment by Guzz-T** [comment]
Unfortunately, a "real" decorator cannot use `self` as an argument. Without `self`, however, the internal fields cannot be accessed correctly. Also `pylint` complains.

**Comment by gerw** [comment]
I tried a little bit and it seems to be possible with a "real" decorator:
````python
def decorator(func):
    def _decorator(self, *args, **kwargs):
        self.num_calls += 1
        return func(self, *args, **kwargs)
    return _decorator

class TestSample:
    def __init__(self):
        self.num_calls = 0

    @decorator
    def dummy(self, foo):
        print(foo)

t = TestSample()

print(f"t.num_calls: {t.num_calls}")
t.dummy(2)
print(f"t.num_calls: {t.num_calls}")
````
There are also no complaints by pylint (only concerning missing docstrings and so on).

**Comment by Bouni** [comment]
@gerw please push a CI trigger commit to check if #134 solved the pytest issue

**Comment by Guzz-T** [comment]
Have you tried accessing private fields like `self._num_calls` instead of `self.num_calls`?

**Comment by gerw** [comment]
Yes, the following is `pylint`-compatible:
```python
"""Random module"""
def decorator(func):
    """Random decorator"""
    def _decorator(self, *args, **kwargs):
        self._num_calls += 1
        return func(self, *args, **kwargs)
    return _decorator

class TestSample:
    """Random class"""
    def __init__(self):
        self._num_calls = 0

    def another_method(self):
        """Random member fctn"""
        return self._num_calls

    @decorator
    def dummy(self, foobar):
        """Random member fctn"""
        print(foobar)

t = TestSample()

print(t.another_method())
t.dummy(2)
print(t.another_method())
```

File: luxtronik/__init__.py [hunk 1]
```diff
@@ -74,43 +74,60 @@ def __del__(self):
...
+                        "Connected to Luxtronik heatpump %s:%s", self._host, self._port
+                    )
+                return func(*args, **kwargs)
+
+        return inner
+
+    def read(self):
+        """
+        Read data from heatpump.
+        All available data will be read from the heat pump.
```
Comment (kbabioch) [inline luxtronik/__init__.py:105]: We should agree on a spelling here, i.e. `heat pump` vs. `heatpump`. Seems like `heat pump` is the "more correct" spelling.

```suggestion
        Read data from heat pump.
        All available data will be read from the heat pump.
```

File: luxtronik/__init__.py [hunk 2]
```diff
@@ -74,43 +74,60 @@ def __del__(self):
...
+
+        def inner(*args, **kwargs):
+            with self._lock:
+                is_none = self._socket is None
+                if is_none:
+                    self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                if is_none or is_socket_closed(self._socket):
+                    self._socket.connect((self._host, self._port))
+                    LOGGER.info(
+                        "Connected to Luxtronik heatpump %s:%s", self._host, self._port
```
Comment (kbabioch) [inline luxtronik/__init__.py:96]: We should agree on a spelling here, i.e. `heat pump` vs. `heatpump`. Seems like `heat pump` is the "more correct" spelling.

```suggestion
                        "Connected to Luxtronik heat pump %s:%s", self._host, self._port
```

File: luxtronik/__init__.py [hunk 3]
```diff
@@ -74,43 +74,60 @@ def __del__(self):
...
+
+    def read(self):
+        """
+        Read data from heatpump.
+        All available data will be read from the heat pump.
+        """
+        return self._connect(self._read)()
+
+    def read_parameters(self):
+        """Read parameters from heatpump."""
```
Comment (kbabioch) [inline luxtronik/__init__.py:110]:
```suggestion
        """Read parameters from heat pump."""
```

File: luxtronik/__init__.py [hunk 4]
```diff
@@ -74,43 +74,60 @@ def __del__(self):
...
+        return self._connect(self._read_parameters)()
+
+    def read_calculations(self):
+        """Read calculations from heatpump."""
```
Comment (kbabioch) [inline luxtronik/__init__.py:114]:
```suggestion
        """Read calculations from heat pump."""
```

File: luxtronik/__init__.py [hunk 5]
```diff
@@ -74,43 +74,60 @@ def __del__(self):
...
+        return self._connect(self._read_calculations)()
+
+    def read_visibilities(self):
+        """Read visibilities from heatpump."""
```
Comment (kbabioch) [inline luxtronik/__init__.py:118]:
```suggestion
        """Read visibilities from heat pump."""
```

File: luxtronik/__init__.py [hunk 6]
```diff
@@ -74,43 +74,60 @@ def __del__(self):
...
+        return self._connect(self._read_visibilities)()
+
+    def write(self, parameters):
+        """
+        Write parameter to heatpump.
```
Comment (kbabioch) [inline luxtronik/__init__.py:123]:
```suggestion
        Write parameter to heat pump.
```

File: luxtronik/__init__.py [hunk 7]
```diff
@@ -74,43 +74,60 @@ def __del__(self):
...
+        All parameters will be written to the heat pump
+        prior to reading back in all data from the heat pump.
         :param Parameters() parameters  Parameter dictionary to be written
                           to the heatpump before reading all available data
-                          from the heatpump. At 'None' it is read only.
+                          from the heatpump.
```
Comment (kbabioch) [inline luxtronik/__init__.py:128]:
```suggestion
                          from the heat pump.
```

**Comment by kbabioch** [comment]
@Guzz-T Could you elaborate on what the vision here is? How do you envision to use this in the future?

**Comment by Guzz-T** [comment]
> 

Does the comment in #130 answer your question?

**Comment by Guzz-T** [comment]
> @Guzz-T Could you elaborate on what the vision here is? How do you envision to use this in the future?

Does the comment in #130 answer your question?

**Comment by Guzz-T** [comment]
> Yes, the following is `pylint`-compatible:

@gerw: I followed your suggestions, but now we get several `W0212 `pylint warnings

**Comment by gerw** [comment]
This is strange. Pylist does not complain about something like `self._num_calls += 1`, but about `self._num_calls = self._num_calls + 1`.

**Comment by Guzz-T** [comment]
> This is strange. Pylist does not complain about something like `self._num_calls += 1`, but about `self._num_calls = self._num_calls + 1`.

I reverted the changes because it would just be a different style of the decorator

**Comment by gerw** [comment]
Since we are not saving the decorated function, it seems not necessary to implement this as a decorator. What about something like
```python
    def _with_lock_and_connect(self, func, *args, **kwargs):
        with self._lock:
            is_none = self._socket is None
            if is_none:
                self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            if is_none or is_socket_closed(self._socket):
                self._socket.connect((self._host, self._port))
                LOGGER.info(
                    "Connected to Luxtronik heat pump %s:%s", self._host, self._port
                )
            return func(*args, **kwargs)


    def read(self):
        return self._with_lock_and_connect(self._read)
```
?

**Comment by Guzz-T** [comment]
There isn't just one way of doing it. My version works like a "@"-decorator.

File: luxtronik/__init__.py [hunk 8]
```diff
@@ -74,43 +88,50 @@ def __del__(self):
...
         Luxtronik controller, which seems unstable otherwise.
-        If write is true, all parameters will be written to the heat pump
-        prior to reading back in all data from the heat pump. If write is
-        false, no data will be written, but all available data will be read
-        from the heat pump.
+        """
+        with self._lock:
+            self._connect()
+            ret_val = func(*args, **kwargs)
+            # self._disconnect()
```
Comment (gerw) [inline luxtronik/__init__.py:103]: This line can be deleted?

File: luxtronik/__init__.py [hunk 9]
```diff
@@ -63,9 +63,23 @@ def __init__(self, host, port=LUXTRONIK_DEFAULT_PORT, safe=True):
         self._port = port
         self._safe = safe
         self._socket = None
-        self.read()
+        self._connect()
 
     def __del__(self):
+        self._disconnect()
+
+    def _connect(self):
```
Comment (gerw) [inline luxtronik/__init__.py:71]: I am not sure, but should 'self._lock' also used in the `_connect` method?

Comment (gerw) [reply luxtronik/__init__.py:71]: (and the same for `_disconnect`?)

File: luxtronik/__init__.py (hunk 8 above)
Comment (Guzz-T) [reply luxtronik/__init__.py:103]: Done. Btw.: It might be better to close the connection afterwards

File: luxtronik/__init__.py (hunk 9 above)
Comment (Guzz-T) [reply luxtronik/__init__.py:71]: `_connect` is called inside `_lock_and_connect` for for multithreaded access to the instance. `__init__` and `__del__` are only called by one thread.

**Review by gerw — APPROVED** [review]

**Pull request merged**